Function Not Send?

Hey there, I wanted to use this in my project, but I'm getting an error that I don't get when manually adjusting the function for recursion:

```

error[E0277]: `(dyn core::future::future::Future<Output = std::result::Result<(subsystems::crawler::IndexCrawler<P>, axiom::actors::Status), std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>>> + 'static)` cannot be sent between threads safely
  --> src/subsystems/crawler.rs:64:50
   |
64 |                     let aid = ctx.system.spawn().with(
   |                                                  ^^^^ `(dyn core::future::future::Future<Output = std::result::Result<(subsystems::crawler::IndexCrawler<P>, axiom::actors::Status), std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>>> + 'static)` cannot be sent between threads safely
   |
   = help: the trait `std::marker::Send` is not implemented for `(dyn core::future::future::Future<Output = std::result::Result<(subsystems::crawler::IndexCrawler<P>, axiom::actors::Status), std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>>> + 'static)`
   = note: required because of the requirements on the impl of `std::marker::Send` for `std::ptr::Unique<(dyn core::future::future::Future<Output = std::result::Result<(subsystems::crawler::IndexCrawler<P>, axiom::actors::Status), std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>>> + 'static)>`
   = note: required because it appears within the type `std::boxed::Box<(dyn core::future::future::Future<Output = std::result::Result<(subsystems::crawler::IndexCrawler<P>, axiom::actors::Status), std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>>> + 'static)>`
   = note: required because it appears within the type `std::pin::Pin<std::boxed::Box<(dyn core::future::future::Future<Output = std::result::Result<(subsystems::crawler::IndexCrawler<P>, axiom::actors::Status), std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>>> + 'static)>>`
   = note: required because of the requirements on the impl of `axiom::actors::Processor<subsystems::crawler::IndexCrawler<P>, std::pin::Pin<std::boxed::Box<(dyn core::future::future::Future<Output = std::result::Result<(subsystems::crawler::IndexCrawler<P>, axiom::actors::Status), std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>>> + 'static)>>>` for `fn(subsystems::crawler::IndexCrawler<P>, axiom::actors::Context, axiom::message::Message) -> std::pin::Pin<std::boxed::Box<(dyn core::future::future::Future<Output = std::result::Result<(subsystems::crawler::IndexCrawler<P>, axiom::actors::Status), std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>>> + 'static)>> {subsystems::crawler::IndexCrawler::<P>::index_crawler}`
```

I noticed a TODO in the code about a `Send` option that is missing, is that related? Would it be simple enough to explain what must happen so I could make a PR?

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "async-recursion"
-version = "0.1.2"
+version = "0.2.0"
 authors = ["Robert Usher <266585+dcchut@users.noreply.github.com>"]
 edition = "2018"
 license = "MIT OR Apache-2.0"

--- a/README.md
+++ b/README.md
@@ -1,8 +1,10 @@
-<h5>Procedural macro for recursive async functions</h5>
+# Async Recursion Macro
+
+Procedural macro for recursive async functions.
 
 Consider the following recursive implementation of the fibonacci numbers:
 
-```rust
+```rust ignore
 async fn fib(n : u32) -> u64 {
    match n {
        0     => panic!("zero is not a valid argument to fib()!"),
@@ -28,7 +30,7 @@ error[E0733]: recursion in an `async fn` requires boxing
 This crate provides an attribute macro to automatically convert an async function
 to one returning a boxed Future.
 
-# Example
+## Example
 
 ```rust
 use async_recursion::async_recursion;
@@ -44,11 +46,22 @@ async fn fib(n : u32) -> u64 {
 }
 ```
 
-# Limitations
-Currently the macro doesn't consider lifetimes at all; this is something I plan to work
-on in the future.
+## ?Send Option
+
+By default the returned future has a `Send` bound to make sure that it can be sent between threads. If this is not desired you can mark that you would like that that bound to be left out like so:
+
+```rust ignore
+#[async_recursion(?Send)]
+async fn example() {}
+```
+
+In other words, `#[async_recursion]` modifies your function to return a [`BoxFuture`] and `#[async_recursion(?Send)]` modifies your function to return a [`LocalBoxFuture`].
+
+[`BoxFuture`]: https://docs.rs/futures/0.3.4/futures/future/type.BoxFuture.html
+[`LocalBoxFuture`]: https://docs.rs/futures/0.3.4/futures/future/type.LocalBoxFuture.html
 
 ### License
+
 Licensed under either of
  * Apache License, Version 2.0
    ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)

--- a/src/expand.rs
+++ b/src/expand.rs
@@ -1,4 +1,4 @@
-use crate::parse::AsyncItem;
+use crate::parse::{AsyncItem, RecursionArgs};
 use proc_macro2::{Span, TokenStream};
 use quote::{quote, ToTokens};
 use syn::punctuated::Punctuated;
@@ -10,8 +10,8 @@ impl ToTokens for AsyncItem {
     }
 }
 
-pub fn expand(item: &mut AsyncItem) {
-    transform_sig(&mut item.0.sig);
+pub fn expand(item: &mut AsyncItem, args: &RecursionArgs) {
+    transform_sig(&mut item.0.sig, &args);
     transform_block(&mut item.0.block);
 }
 
@@ -42,7 +42,7 @@ impl ArgLifetime {
 //
 // Output:
 //     fn f<S, T>(x : S, y : &T) -> Pin<Box<dyn Future<Output = Ret> + Send>
-fn transform_sig(sig: &mut Signature) {
+fn transform_sig(sig: &mut Signature, args: &RecursionArgs) {
     // Determine the original return type
     let ret = match &sig.output {
         ReturnType::Default => quote!(()),
@@ -157,11 +157,15 @@ fn transform_sig(sig: &mut Signature) {
         box_lifetime.push(asr);
     }
 
-    // TODO: we should have an attribute for whether we want a Send bound here or not
+    let send_bound: TokenStream = if args.send_bound {
+        quote!(+ ::core::marker::Send)
+    } else {
+        quote!()
+    };
 
     // Modify the return type
     sig.output = parse_quote! {
         -> core::pin::Pin<Box<
-            dyn core::future::Future<Output = #ret> #(+ #box_lifetime)*>>
+            dyn core::future::Future<Output = #ret> #(+ #box_lifetime)* #send_bound >>
     };
 }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1,8 +1,10 @@
-//! <h5>Procedural macro for recursive async functions</h5>
+//! # Async Recursion Macro
+//!
+//! Procedural macro for recursive async functions.
 //!
 //! Consider the following recursive implementation of the fibonacci numbers:
 //!
-//! ```compile_fail
+//! ```rust ignore
 //! async fn fib(n : u32) -> u64 {
 //!    match n {
 //!        0     => panic!("zero is not a valid argument to fib()!"),
@@ -18,19 +20,19 @@
 //! ```console
 //! error[E0733]: recursion in an `async fn` requires boxing
 //! --> src/main.rs:1:26
-//! |
+//!  |
 //! 1 | async fn fib(n : u32) -> u64 {
-//! |                          ^^^ recursive `async fn`
-//! |
-//! = note: a recursive `async fn` must be rewritten to return a boxed `dyn Future`.
+//!  |                          ^^^ recursive `async fn`
+//!  |
+//!  = note: a recursive `async fn` must be rewritten to return a boxed `dyn Future`.
 //! ```
 //!
 //! This crate provides an attribute macro to automatically convert an async function
 //! to one returning a boxed Future.
 //!
-//! # Example
+//! ## Example
 //!
-//! ```
+//! ```rust
 //! use async_recursion::async_recursion;
 //!
 //! #[async_recursion]
@@ -44,16 +46,27 @@
 //! }
 //! ```
 //!
-//! # Limitations
-//! Currently the macro doesn't consider lifetimes at all; this is something I plan to work
-//! on in the future.
+//! ## ?Send Option
+//!
+//! By default the returned future has a `Send` bound to make sure that it can be sent between threads. If this is not desired you can mark that you would like that that bound to be left out like so:
+//!
+//! ```rust ignore
+//! #[async_recursion(?Send)]
+//! async fn example() {}
+//! ```
+//!
+//! In other words, `#[async_recursion]` modifies your function to return a [`BoxFuture`] and `#[async_recursion(?Send)]` modifies your function to return a [`LocalBoxFuture`].
+//!
+//! [`BoxFuture`]: https://docs.rs/futures/0.3.4/futures/future/type.BoxFuture.html
+//! [`LocalBoxFuture`]: https://docs.rs/futures/0.3.4/futures/future/type.LocalBoxFuture.html
 //!
 //! ### License
+//!
 //! Licensed under either of
-//! * Apache License, Version 2.0
-//!   ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
-//! * MIT license
-//!   ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)
+//!  * Apache License, Version 2.0
+//!    ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
+//!  * MIT license
+//!    ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)
 //! at your option.
 
 extern crate proc_macro;
@@ -62,17 +75,18 @@ mod expand;
 mod parse;
 
 use crate::expand::expand;
-use crate::parse::AsyncItem;
+use crate::parse::{AsyncItem, RecursionArgs};
 
 use proc_macro::TokenStream;
 use quote::quote;
 use syn::parse_macro_input;
 
 #[proc_macro_attribute]
-pub fn async_recursion(_args: TokenStream, input: TokenStream) -> TokenStream {
+pub fn async_recursion(args: TokenStream, input: TokenStream) -> TokenStream {
     let mut item = parse_macro_input!(input as AsyncItem);
+    let args = parse_macro_input!(args as RecursionArgs);
 
-    expand(&mut item);
+    expand(&mut item, &args);
 
     TokenStream::from(quote!(#item))
 }

--- a/src/parse.rs
+++ b/src/parse.rs
@@ -1,6 +1,6 @@
 use proc_macro2::Span;
 use syn::parse::{Error, Parse, ParseStream, Result};
-use syn::ItemFn;
+use syn::{token::Question, ItemFn, Token};
 
 pub struct AsyncItem(pub ItemFn);
 
@@ -14,5 +14,35 @@ impl Parse for AsyncItem {
         }
 
         Ok(AsyncItem(item))
+    }
+}
+
+pub struct RecursionArgs {
+    pub send_bound: bool,
+}
+
+impl Default for RecursionArgs {
+    fn default() -> Self {
+        RecursionArgs { send_bound: true }
+    }
+}
+
+/// Custom keywords for parser
+mod kw {
+    syn::custom_keyword!(Send);
+}
+
+impl Parse for RecursionArgs {
+    fn parse(input: ParseStream) -> Result<Self> {
+        // Check for the `?Send` option
+        if input.peek(Token![?]) {
+            input.parse::<Question>()?;
+            input.parse::<kw::Send>()?;
+            Ok(Self { send_bound: false })
+        } else if !input.is_empty() {
+            return Err(input.error("expected `?Send` or empty"));
+        } else {
+            Ok(Self::default())
+        }
     }
 }

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -27,7 +27,9 @@ impl<T> Node<'_, T> {
     }
 }
 
-#[async_recursion]
+// Note: Use the `?Send` notation here alows us not to require that our type parameter `T` is
+// `T: PartialEq + Sync + Send`.
+#[async_recursion(?Send)]
 async fn contains_value<'a, T>(value: &T, node: &Node<'a, T>) -> bool
 where
     T: PartialEq,
@@ -40,7 +42,7 @@ where
     }
 }
 
-#[async_recursion]
+#[async_recursion(?Send)]
 async fn contains_value_2<'a, 'b, T>(value: &'b T, node: &'b Node<'a, T>) -> bool
 where
     T: PartialEq,
@@ -113,4 +115,16 @@ fn struct_method_fib() {
         assert_eq!(e.fib(5).await, 5);
         assert_eq!(e.fib(7).await, 13);
     });
+}
+
+#[test]
+fn future_is_send() {
+    let fut = fib(6);
+
+    let child = std::thread::spawn(move || {
+        let result = block_on(fut);
+        assert_eq!(result, 8);
+    });
+
+    child.join().unwrap();
 }

